### PR TITLE
[RPA v3] Fix int16 overflow in position computation for sequences > 32K tokens

### LINKED ISSUE #2755: [Bug] int16 overflow in RPA v3 position computation causes incorrect attention for sequences > 32K tokens
## Bug Description

The RPA v3 decode kernel uses `int16` for position indices on TPU v6+/v7 with packed dtypes (bf16/fp16), which overflows for sequences longer than 32,767 tokens. This causes incorrect causal and sliding window attention masks during decode, producing wrong model outputs.

## Location

`tpu_inference/kernels/ragged_paged_attention/v3/kernel.py`, lines 469-471:

```python
int_ty = jnp.int32
if get_dtype_packing(q_dtype) != 1 and get_tpu_version() >= 6:
    int_ty = jnp.int16  # overflows for seq_len > 32767
```

## Impact

- Affects ALL models with sliding window attention on TPU v6+/v7 using bf16/fp16 when `max_model_len > 32768`
- Gemma 3 (sliding_window=1024) and Gemma 4 (sliding_window=512) are both affected
- All Gemma models support 128K-256K context — well beyond the int16 limit
- Models without sliding window (e.g., Qwen3) are not affected because the sliding window mask path is not triggered

## Symptoms

- Severe recency bias in generated text: model disproportionately attends to the end of the prompt
- Position 37,500 (int16) wraps to -28,036, corrupting both `q_span >= k_span` (causal mask) and `q_span < k_span + sliding_window` (sliding window mask)
- Output quality degrades significantly compared to GPU for the same model and input when prompts exceed 32K tokens

## Reproduction

1. Deploy a model with sliding window attention (e.g., Gemma 3 4B IT or Gemma 4 E4B IT) on TPU v7 with `--max-model-len=65536 --dtype=bfloat16`
2. Send a prompt with >32K tokens containing structured content (e.g., a long numbered list)
3. Observe that the model's output disproportionately references items near the end of the list
4. Compare with GPU output which correctly references items across the full list

## Validated Fix

Guard the int16 optimization with a max sequence length check:

```python
int_ty = jnp.int32
max_kv_len = pages_per_seq * page_size
if (get_dtype_packing(q_dtype) != 1 and get_tpu_version() >= 6
        and max_kv_len <= 32767):
    int_ty = jnp.int16
```

Validated on:
- **Gemma 3 4B IT**: TPU output distribution matches GPU after fix
- **Gemma 4 E4B IT**: TPU output distribution matches GPU after fix

## Environment

- TPU: v7x (4 chips, topology 2x2x1)
- Image: `vllm/vllm-tpu:nightly` (v0.21.1rc1.dev352)
- Models tested: `google/gemma-3-4b-it`, `google/gemma-4-E4B-it`

### PR DESCRIPTION
**Pull request by ziy**
## Summary

Fix int16 overflow in RPA v3 decode kernel position computation that causes incorrect attention masks for sequences longer than 32,767 tokens.

The kernel uses `int16` for position indices as a VMEM optimization on TPU v6+/v7 with packed dtypes (bf16/fp16). However, `int16` overflows for sequences beyond 32,767 tokens, corrupting both the causal mask (`q_span >= k_span`) and sliding window mask (`q_span < k_span + sliding_window`). This produces severe recency bias for models with sliding window attention (Gemma 3, Gemma 4) when serving long prompts.

## Fix

Guard the `int16` optimization with a max sequence length check (`pages_per_seq * page_size <= 32767`). For longer sequences, `int32` is used — correct but slightly less VMEM-efficient.

## Test plan

Validated on TPU v7x (4 chips, `vllm/vllm-tpu:nightly`):

| Model | Output distribution (before fix) | Output distribution (after fix) | GPU reference |
|---|---|---|---|
| Gemma 3 4B IT | 71% concentrated at end of prompt | Matches GPU distribution | ✅ |
| Gemma 4 E4B IT | 95% concentrated at end of prompt | Matches GPU distribution | ✅ |
| Qwen3 4B (no sliding window) | Correct (unaffected) | Correct | ✅ |

Benchmark: Long prompts (~37K tokens) with structured content. Prompts exceed int16 max (32,767), triggering the overflow. After fix, TPU and GPU produce matching output distributions for all tested models.

Fixes #2755

### CODE REVIEW

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 1]
```diff
@@ -467,7 +467,9 @@ def flash_attention_step1_qk_softmax(
             s = soft_cap * jnp.tanh(s / soft_cap)
 
         int_ty = jnp.int32
-        if get_dtype_packing(q_dtype) != 1 and get_tpu_version() >= 6:
+        max_kv_len = pages_per_seq * page_size
+        if (get_dtype_packing(q_dtype) != 1 and get_tpu_version() >= 6
+                and max_kv_len <= 32767):
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: very very nit: instead of 32767, maybe something like `jnp.iinfo(jnp.int16).max` is more intuitively understood?

Comment (ziy) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: Done — updated to use `jnp.iinfo(jnp.int16).max` instead of the magic number. Thanks for the suggestion!

**Comment by ziy** [comment]
Gentle ping on this one 🙏 — it's a correctness fix for int16 position overflow on >32K-token sequences (causes recency bias on Gemma 3/4 SWA models). @kyuyeunk's nit is addressed (now using `jnp.iinfo(jnp.int16).max`). CI shows a `pipeline-upload` failure, which I believe is the fork-PR gate rather than a real test failure — could a maintainer add the `ready` label to trigger CI, and take a look when you have a moment? @bythew3i @a1yssan13 @jrplatin. Thanks!

**Review by kyuyeunk — APPROVED** [review]

**Comment by kyuyeunk** [comment]
Approved!
